runtime/src/main/java/org/corfudb/runtime/object/transactions/SnapshotTransactionalContext.java

```diff
@@ -39,15 +39,22 @@ public <R, T extends ICorfuSMR<T>> R access(ICorfuSMRProxyInternal<T> proxy,
                         && !o.isOptimisticallyModifiedUnsafe(),
                 o -> syncWithRetryUnsafe(o, getSnapshotTimestamp(), proxy, null),
                 accessFunction::access,
-                version -> updateKnownStreamPosition(proxy.getStreamID(), version));
+                version -> updateKnownStreamPosition(proxy, version));
     }
 
     /**
      * {@inheritDoc}
      */
     @Override
     public long commitTransaction() throws TransactionAbortedException {
-        return getMaxAddressRead();
+        // If the transaction has read monotonic objects, we instead return the min address
+        // of all accessed streams. Although this avoids data loss, clients subscribing at
+        // this point for delta/streaming may observe duplicate data.
+        if (hasAccessedMonotonicObject) {
+            return getMinAddressRead();
+        } else {
+            return getMaxAddressRead();
+        }
```

Comment:
What if the tables were empty and this code path is triggered 

```

            if (commitAddress == Address.NON_ADDRESS) {
                // If no streams were touched or only empty streams were touched, the returned address would be -1
                // But -1 can be detrimental to stream subscription which is just looking for a safe spot
                // to resume subscription from, so instead return the address from the snapshot token taken.
                long snapshotAddress = context.getSnapshotTimestamp().getSequence();
                if (snapshotAddress >= 0) {
                    // Perform a read at this address to force materialization.
                    runtime.getAddressSpaceView().read(snapshotAddress);
                }
                return snapshotAddress;
            }
```

Then data loss can occur when the sequencer regresses before `read(snapshotAddress)` operation occurs. Imagine snapshotAddress = 10 and the sequencer regresses to version 5, the read will materialize 10, but all updates from 5 to 10 won't be delivered. 

Your code is node introducing this issue, but it's relying on existing code that is incorrect.